### PR TITLE
Initialize a global tracer with opentracing and jaegar in the service module

### PR DESCRIPTION
**Pull request by madhuravi**

### CODE REVIEW

**Comment by madhuravi** [comment]
@uber-go/service-framework

File: core/tracing/tracer_test.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func TestLoadAppConfig(t *testing.T) {
+	jConfig := loadAppConfig(emptyJaegerConfig, logger)
+	assert.NotNil(t, jConfig)
+	assert.NotNil(t, jConfig.Logger)
+}
+
+func TestLoadAppConfig_JaegerConfigWithLogger(t *testing.T) {
+	jConfig := loadAppConfig(jaegerConfigWithLogger, logger)
```
Comment (anuptalwalkar) [inline core/tracing/tracer_test.go]: nit: may be could use helper withjaegerConfig func wrap up for the tests.

File: service/service.go [hunk 2]
```diff
@@ -138,6 +139,20 @@ func New(options ...Option) (Owner, error) {
 		metrics.Freeze()
 	}
 
+	if svc.Tracer() == nil {
+		svc.configProvider.GetValue("tracing").PopulateStruct(&svc.tracerConfig)
```
Comment (anuptalwalkar) [inline service/service.go]: Not sure how lint passed, but you need to check for err returned on PopulateStruct

File: core/tracing/tracer.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+)
+
+// InitGlobalTracer instantiates a new global tracer
+func InitGlobalTracer(
+	cfg *jaegerconfig.Configuration,
+	serviceName string,
+	logger ulog.Log,
+	scope tally.Scope,
+) (opentracing.Tracer, error) {
+	appCfg := cfg
```
Comment (anuptalwalkar) [inline core/tracing/tracer.go]: Why not just use cfg?

File: core/tracing/tracer.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+type jaegerLogger struct {
+	log ulog.Log
+}
+
+func (jl *jaegerLogger) Error(msg string) {
+	jl.log.Error(msg)
+}
+
+func (jl *jaegerLogger) Infof(msg string, args ...interface{}) {
```
Comment (anuptalwalkar) [inline core/tracing/tracer.go]: comment on exported methods?

**Comment by sectioneight** [comment]
I disabled errcheck on _test files since it's kinda annoying and not that
helpful

File: core/tracing/tracer.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	"go.uber.org/fx/core/ulog"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber-go/tally"
+	jaegerconfig "github.com/uber/jaeger-client-go/config"
+)
+
+// InitGlobalTracer instantiates a new global tracer
+func InitGlobalTracer(
+	cfg *jaegerconfig.Configuration,
```
Comment (madhuravi) [inline core/tracing/tracer.go]: Keeping this consistent with other core components like logging. Check service.go

File: core/tracing/tracer.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		appCfg = loadAppConfig(cfg, logger)
+		reporter = &jaegerReporter{
+			reporter: scope.Reporter(),
+		}
+	}
+	tracer, closer, err := appCfg.New(serviceName, reporter)
+	if err != nil {
+		return tracer, err
+	}
+	defer closer.Close()
```
Comment (madhuravi) [inline core/tracing/tracer.go]: This actually needs to be in service shutdown, moving it there.

File: core/tracing/tracer_test.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/uber-go/tally"
+	"github.com/uber-go/zap"
+	"github.com/uber/jaeger-client-go"
+	jaegerconfig "github.com/uber/jaeger-client-go/config"
+)
+
+var (
+	serviceName            = "serviceName"
```
Comment (madhuravi) [inline core/tracing/tracer_test.go]: are you saying this should be renamed to "_serviceName"? Don't see this naming elsewhere in the code.

File: core/tracing/tracer_test.go (hunk 1 above)
Comment (madhuravi) [inline core/tracing/tracer_test.go]: Don't see that making these tests simpler/cleaner. Tests are simple enough and use different combinations of input arguments each time.

File: service/service.go (hunk 2 above)
Comment (madhuravi) [inline service/service.go]: Good catch.

File: core/tracing/tracer.go (hunk 4 above)
Comment (madhuravi) [inline core/tracing/tracer.go]: Done.

File: core/tracing/tracer.go (hunk 3 above)
Comment (madhuravi) [inline core/tracing/tracer.go]: Done.

File: core/tracing/tracer.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
+package tracing
+
+import (
+	"time"
+
+	"go.uber.org/fx/core/ulog"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber-go/tally"
+	jaegerconfig "github.com/uber/jaeger-client-go/config"
```
Comment (sectioneight) [inline core/tracing/tracer.go]: why alias this? generally we should only alias a package if it conflicts with another import, but we're not importing another `config` package here, right?

File: core/tracing/tracer.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	var reporter *jaegerReporter
+	if cfg == nil || !cfg.Disabled {
+		appCfg = loadAppConfig(cfg, logger)
+		reporter = &jaegerReporter{
+			reporter: scope.Reporter(),
+		}
+	}
+	tracer, closer, err := appCfg.New(serviceName, reporter)
+	if err != nil {
+		return tracer, err
```
Comment (sectioneight) [inline core/tracing/tracer.go]: use `errors.Wrap` (search for other examples) whenever bubbling up errors from now on (i just added this today, so it's a new pattern)

File: core/tracing/tracer.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		jaegerlogger := &jaegerLogger{
+			log: logger,
+		}
+		appCfg.Logger = jaegerlogger
+	}
+	return appCfg
+}
+
+type jaegerLogger struct {
+	log ulog.Log
```
Comment (sectioneight) [inline core/tracing/tracer.go]: why not just embed this directly? then you don't have to proxy Error, Infof, etc.

File: core/tracing/tracer_test.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"testing"
+
+	"go.uber.org/fx/core/ulog"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/uber-go/tally"
+	"github.com/uber-go/zap"
+	"github.com/uber/jaeger-client-go"
+	jaegerconfig "github.com/uber/jaeger-client-go/config"
```
Comment (sectioneight) [inline core/tracing/tracer_test.go]: just use the natural package name, config

File: core/tracing/tracer_test.go (hunk 7 above)
Comment (sectioneight) [inline core/tracing/tracer_test.go]: yes, all globals at uber should be underscore-prefixed. we may have missed a few from shawn's initial prototype, but anup is correct that going forward we should fix these to conform to our style

File: core/tracing/tracer_test.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+)
+
+var (
+	serviceName            = "serviceName"
+	logger                 = ulog.Logger()
```
Comment (sectioneight) [inline core/tracing/tracer_test.go]: grabbing this is a global seems weird. can you make a function that returns a ulog.Logger and use that instead? otherwise we get into weird situations where people configure the logger but you've got a reference to an older version. just generally avoid globals as much as possible, even in tests

File: core/ulog/in_memory_log.go [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+	"github.com/uber-go/zap"
+)
+
+// TestBuffer is a buffer used to test the zap logger
+type TestBuffer struct {
+	bytes.Buffer
+}
+
+// Sync is a noop
```
Comment (sectioneight) [inline core/ulog/in_memory_log.go]: maybe mention that it's here to satisfy some interface?

File: core/ulog/in_memory_log.go [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	return output[:len(output)-1]
+}
+
+// Stripped returns buffer as a string without the newline
+func (b *TestBuffer) Stripped() string {
+	return strings.TrimRight(b.String(), "\n")
+}
+
+// WithInMemoryLogger creates an in-memory zap logger that can be used in tests
+func WithInMemoryLogger(t *testing.T, opts []zap.Option, f func(zap.Logger, *TestBuffer)) {
```
Comment (sectioneight) [inline core/ulog/in_memory_log.go]: why is this exported? are you using it outside of this package? if not, please un-export. exported tests methods are a bit awkward (and yes, i've written a few, but very intentionally).

File: service/service.go [hunk 15]
```diff
@@ -138,6 +139,20 @@ func New(options ...Option) (Owner, error) {
...
+		tracer, err := tracing.InitGlobalTracer(
+			&svc.tracerConfig,
+			svc.standardConfig.ServiceName,
+			svc.log,
+			svc.scope.SubScope("tracing"),
+		)
+		if err != nil {
+			return svc, err
```
Comment (sectioneight) [inline service/service.go]: wrap error

File: core/tracing/tracer_test.go (hunk 7 above)
Comment (madhuravi) [inline core/tracing/tracer_test.go]: This is not a global though. Is this specifically for package-scope variables? Can you send me a few examples? Not able to find any code samples for this in uberfx or opengrok.

Comment (sectioneight) [inline core/tracing/tracer_test.go]: Anything that's not scoped to a function or a method is a global.

https://github.com/uber-go/fx/blob/master/core/config/config.go#L44-L50

Again, we haven't fixed them all, but this is the standard.

File: core/ulog/in_memory_log.go (hunk 14 above)
Comment (madhuravi) [inline core/ulog/in_memory_log.go]: yes, using this is ulog_test and tracer_test.

File: core/tracing/tracer_test.go (hunk 12 above)
Comment (madhuravi) [inline core/tracing/tracer_test.go]: Good point. It does make tests simpler/more compact though.

**Comment by madhuravi** [comment]
@sectioneight Seeing this lint error in the travis build (though it succeeds locally), any idea what it is? It has a reference to your personal repo but I dont see a dependency on this in uberfx glide. 

# github.com/sectioneight/md-to-godoc/render
../../github.com/sectioneight/md-to-godoc/render/godoc.go:30: undefined: blackfriday.CommonExtensions
../../github.com/sectioneight/md-to-godoc/render/godoc.go:42: cannot use GodocRenderer literal (type *GodocRenderer) as type blackfriday.Renderer in return argument:
	*GodocRenderer does not implement blackfriday.Renderer (missing AutoLink method)

File: core/tracing/tracer.go (hunk 10 above)
Comment (madhuravi) [inline core/tracing/tracer.go]: The jaeger logger has a different interface, this is a conversion layer from jaeger logger to uberfx logger interface.

**Comment by sectioneight** [comment]
Hmm i just merged some changes to the markdown tool, will look tomorrow.
Sorry for the interruption, it's almost certainly my fault.

File: core/ulog/in_memory_log.go (hunk 14 above)
Comment (anuptalwalkar) [inline core/ulog/in_memory_log.go]: We can move this to testutil at some point and use in-memory logger for validating logging messages. I am ok with exporting for now, may be add a TODO?

Comment (madhuravi) [inline core/ulog/in_memory_log.go]: Good idea, I wasn't aware of testutils. I'll just move it there now.

**Comment by sectioneight** [comment]
Yea, testutils seems like the right place (as long as we're conscious not
to make it a dumping ground). Fine to do in a future PR though.

File: service/host.go [hunk 16]
```diff
@@ -149,6 +149,13 @@ func (s *host) shutdown(err error, reason string, exitCode *int) (bool, error) {
 		s.scope.Close()
 	}
 
+	// Flush tracing buffers
+	if s.tracerCloser != nil {
+		if err = s.tracerCloser.Close(); err != nil {
+			ulog.Logger().Error("Failure to close tracer", "error", err)
```
Comment (sectioneight) [inline service/host.go]: This is the global logger. This should use `s.Logger()` instead of the global one (there's also a mistake above which is probably why you did this, you don't need to necessarily fix that one here but we should make a note to do it later)

File: service/host.go [hunk 17]
```diff
@@ -149,6 +149,13 @@ func (s *host) shutdown(err error, reason string, exitCode *int) (bool, error) {
 		s.scope.Close()
 	}
 
+	// Flush tracing buffers
+	if s.tracerCloser != nil {
```
Comment (sectioneight) [inline service/host.go]: Let's log.Debug that we're closing the tracer

File: core/tracing/tracer.go [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+		jaegerlogger := &jaegerLogger{
+			log: logger,
+		}
+		cfg.Logger = jaegerlogger
+	}
+	return cfg
+}
+
+type jaegerLogger struct {
+	log ulog.Log
```
Comment (sectioneight) [inline core/tracing/tracer.go]: I tried to leave this before, but why not embed `log` directly, then you don't have to write the `Error` and `Infof` proxy methods yourself below

File: core/tracing/tracer_test.go [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+func TestInitGlobalTracer_Disabled(t *testing.T) {
+	tracer, closer, err := InitGlobalTracer(_disabledJaegerConfig, _serviceName, getLogger(), _scope)
+	assert.NotNil(t, tracer)
+	assert.NotNil(t, closer)
+	assert.NoError(t, err)
+}
+
+func TestInitGlobalTracer_NoServiceName(t *testing.T) {
+	tracer, closer, err := InitGlobalTracer(_emptyJaegerConfig, "", getLogger(), _scope)
+	assert.NotNil(t, err)
```
Comment (sectioneight) [inline core/tracing/tracer_test.go]: `assert.Error(t, err)`

File: modules/rpc/yarpc_test.go [hunk 20]
```diff
@@ -22,11 +22,11 @@ package rpc
 
 import (
 	"bytes"
-	"context"
 	"testing"
 
 	"go.uber.org/yarpc/encoding/raw"
 	"go.uber.org/yarpc/transport"
+	"golang.org/x/net/context"
```
Comment (sectioneight) [inline modules/rpc/yarpc_test.go]: Why this change? Did you accidentally downgrade yarpc? Newer versions of YARPC take the stdlib context.

File: service/host.go [hunk 21]
```diff
@@ -140,7 +140,7 @@ func (s *host) shutdown(err error, reason string, exitCode *int) (bool, error) {
 	errs := s.stopModules()
 	if len(errs) > 0 {
 		for k, v := range errs {
-			ulog.Logger().Error("Failure to shut down module", "name", k.Name(), "error", v.Error())
+			s.Logger().Error("Failure to shut down module", "name", k.Name(), "error", v.Error())
```
Comment (sectioneight) [inline service/host.go]: noice

File: core/tracing/tracer.go (hunk 18 above)
Comment (glibsm) [inline core/tracing/tracer.go]: @madhuravi I think this comment is good. We should change `log ulog.Log` to just `ulog.Log`, then you can delete a bunch of forwarding methods from here

Comment (madhuravi) [inline core/tracing/tracer.go]: I spoke to Aiden about this offline and forgot to update the comment. The methods signatures are not the same, see "Infof" versus "Info.
